attributes-next.md

```diff
@@ -0,0 +1,252 @@
+---
+RFC: unassigned
+Title: Attributes v.Next
+Author: Lamont Granquist <lamont@chef.io>
+Status: Draft
+Type: <Standards Track>
+---
+
+# Attributes v.Next
+
+Overhauling Attributes to be simpler.
+
+## Motivation
+
+    As a Chef User,
+    I want Attributes to be Less Confusing.
+
+    As a Chef User,
+    I dislike NoMethodError for NilClass Errors.
+
+    As a Chef Developer,
+    The monolithic Node attributes API tries to do too much with a shared namespace.
+
+    As a Maintainer of Existing Cookbooks,
+    I need a slow and backwards compatibile transition to any new API.
+
+    As a Ohai Plugin Developer,
+    I want to be able to add ohai attributes without clobbering someone's cookbook attributes.
+
+    As a Chef Support Person,
+    We see normal attribute persistence cause consistent user confusion.
+
+    As a Chef Support Person,
+    We see array merging (sometimes) to cause consistent user confusion, with most users being surprised by merging behavior.
+
+    As a Chef Support Person,
+    We see consistent user confusion over what the point of force_default is.
+
+    As a Chef Developer,
+    I've never seen anyone use or understand the role/environment precedence order swap between default and override.
+
+## Overall Design
+
+Migration Strategy:
+
+- Existing APIs remain unchanged
+- Deprecation warnings will be emitted for edge cases where the APIs differ
+- Ater fixing the deprecation warnings, autofixing from the old APIs to the new APIs should be possible
+
+Three Separated Namespaces:
+
+- Ohai attributes
+- Default and override attributes
+- Normal/storage attributes
+
+Deprecated Behavior:
+
+- Array deep merging is removed (must be manually replaced with hash-key merging)
```

Comment:
It already does.   Which is why this shit is so confusing.

Array merging only happens between "same" precedence levels -- default to default, or override to override:

attributes/default.rb:
```
default["foo"]["bar"] = [ 80 ]
```

recipes/default.rb:
```
default["foo"]["bar"] = [ 8080 ]
pp node["foo"]["bar"] #  => [ 80, 8080 ]
force_default["foo"]["bar"] = [ 8181 ]
pp node["foo"]["bar"] #  => [ 80, 8080, 8181 ]
node.override["foo"]["bar"] = [ 25 ]
pp node["foo"]["bar"] #  => [ 25 ]
```